README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
'fedregs': Text Analysis of the US Code of Federal Regulations
--------------------------------------------------------------

[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active) [![codecov](https://codecov.io/gh/slarge/fedregs/branch/master/graph/badge.svg)](https://codecov.io/gh/slarge/fedregs) [![Travis-CI Build Status](https://travis-ci.org/slarge/fedregs.svg?branch=master)](https://travis-ci.org/slarge/fedregs) <!-- [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/fedregs)](https://cran.r-project.org/package=fedregs) --> <!-- [![downloads](http://cranlogs.r-pkg.org/badges/grand-total/fedregs) -->

The goal of `fedregs` is to allow for easy exploration and analysis of the [Code of Federal Regulation](https://www.gpo.gov/fdsys/browse/collectionCfr.action?selectedYearFrom=2017&go=Go).

Installation
------------

You can install `fedregs` using:

``` r
install.packages("fedregs")
# Or: devtools::install_github("slarge/fedregs")
```

Example
-------

The [Code of Federal Regulation](https://www.gpo.gov/help/index.html#about_code_of_federal_regulations.htm) is organized according to a consistent hierarchy: title, chapter, part, subpart, section, and subsection. Each title within the CFR is (somewhat haphazardly) divided into volumes and over time each chapter isn't consistently in the same volume. The `cfr_text()` function is the main function in the package and it will return the text for a specified part, including the associated subparts and sections. Behind the scenes, `cfr_text()` and associated helper functions gather the volumes for a given title/year combination and parses XML to determine the chapters, parts, and subparts associated with each volume. Next, the text is extracted for each subpart. The `return_tidytext = TRUE` argument will return a tibble with the text in a [tidytext](https://www.tidytextmining.com/tidytext.html) format. If *ngrams* are your game, set `token = "ngrams"` and specify `n`.

``` r
library(fedregs)
library(dplyr)
library(tidyr)
library(ggplot2)
library(quanteda)

regs <- cfr_text(year = 2017,
                 title_number = 50,
                 chapter = 6,
                 part = 648,
                 #token = "ngrams", # uncomment for ngrams of length 2
                 #n = 2, # uncomment for ngrams of length 2
                 return_tidytext = TRUE,
                 verbose = FALSE)
head(regs)
## # A tibble: 6 x 6
## # Groups:   subpart, year, title_number, chapter, part [6]
##   subpart                      year title_number chapter  part         data
##   <chr>                       <dbl>        <dbl> <chr>   <dbl> <list<df[,4>
## 1 Subpart A—General Provisio~  2017           50 VI        648 [84,596 x 4]
## 2 Subpart B—Management Measu~  2017           50 VI        648 [10,384 x 4]
## 3 Subpart C—Management Measu~  2017           50 VI        648    [696 x 4]
## 4 Subpart D—Management Measu~  2017           50 VI        648 [25,567 x 4]
## 5 Subpart E—Management Measu~  2017           50 VI        648  [6,994 x 4]
## 6 Subpart F—Management Measu~  2017           50 VI        648 [97,477 x 4]
```

Now, we can unnest the tibble and take a peek at the data to see what data we have to play with.

``` r
regs %>%
  unnest(cols = c(data)) %>% head(20) %>% pull(word)
##  [1] "c"          "a"          "this"       "part"       "implements"
##  [6] "the"        "fishery"    "management" "plans"      "fmps"      
## [11] "for"        "the"        "atlantic"   "mackerel"   "squid"     
## [16] "and"        "butterfish" "fisheries"  "atlantic"   "mackerel"
```

Not entirely unexpected, but there are quite a few common words that don't mean anything. These "stop words" typically don't have important significance and and are filtered out from search queries.

``` r
head(stopwords("english"))
## [1] "i"      "me"     "my"     "myself" "we"     "our"
```

There are some other messes like punctuation, numbers, *i*ths, Roman Numerals, web sites, and random letters (probably from indexed lists) that can be removed with some simple regex-ing. We can also convert the raw words to word stems to further aggregate our data.

``` r
stop_words <- data_frame(word = stopwords("english"))
## Warning: `data_frame()` is deprecated, use `tibble()`.
## This warning is displayed once per session.

clean_words <- regs %>%
  unnest(cols = c(data)) %>% 
  mutate(word = gsub("[[:punct:]]", "", word), # remove any remaining punctuation
                word = gsub("^[[:digit:]]*", "", word)) %>%  # remove digits (e.g., 1st, 1881a, 15th, etc)
  anti_join(stop_words, by = "word") %>%  # remove "stop words"
  filter(is.na(as.numeric(word)),
                !grepl("^m{0,4}(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$",
                      word), # adios Roman Numerals
                !grepl("\\b[a-z]{1}\\b", word), # get rid of one letter words
                !grepl("\\bwww*.", word)) %>% # get rid of web addresses
  mutate(word = tokens(word),
                word = as.character(tokens_wordstem(word)))
## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion
## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion

## Warning in ~is.na(as.numeric(word)): NAs introduced by coercion
head(clean_words)
## # A tibble: 6 x 9
## # Groups:   subpart, year, title_number, chapter, part [1]
##   subpart  year title_number chapter  part SECTION_NAME SECTION_NUMBER
##   <chr>   <dbl>        <dbl> <chr>   <dbl> <chr>        <chr>         
## 1 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## 2 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## 3 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## 4 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## 5 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## 6 Subpar~  2017           50 VI        648 Purpose and~ §<U+2009>648.1       
## # ... with 2 more variables: values <chr>, word <chr>
```

Now we can look at binning and plotting the words

``` r
count_words <- clean_words %>%
  group_by(word) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  arrange(-n) %>% 
  top_n(n = 50, wt = n) %>% 
  mutate(word = reorder(word, n))
```

``` r
ggplot(count_words, aes(word, n)) +
  geom_col() +
  labs(xlab = NULL, 
       title = "Code of Federal Regulations", 
       subtitle = "Title 50, Chapter VI, Part 648",
       caption = sprintf("Data accessed on %s from:\n https://www.gpo.gov/fdsys/browse/collectionCfr.action?collectionCode=CFR", 
                         format(Sys.Date(), "%d %B %Y"))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom",
        text = element_text(size = 8)) +
  coord_flip() +
  theme_minimal()
```

<img src="README_figs/README-plot_words-1.png" width="672" />


NEWS.md
## fedregs 1.0.0

* Big change that might hinder backwards compatability, so it gets a new big number.
* Update code to work with tidyr >= 1.0.0.
* Update code to download xml to temporary directory instead of repeatedly querying the API servers.
* Depricate cfr_url()

## fedregs 0.1.1

* Updated code to work with the new Code of Federal Regulations API
* Maximum year increased from 2017 to 2018
* Added improved error checks in case XML format is off 

# fedregs 0.1.0

## New Stuff

* Initial release.

tests/testthat/test-fedregs.R
context("test-fedregs.R")

# test_that("We get the best CFR URLs.", {
#
#   good_year <- 2000
#   good_title_number <- 15
#
#   bad_year <- 1993
#   bad_title_number <- 55
#
#   na_year <- 1996
#   na_title_number <- 1
#
#   testthat::expect_error(cfr_urls(year = bad_year,
#                           title_number = good_title_number,
#                             check_url = TRUE), "Year must be between 1996 and 2018.\n")
#
#   testthat::expect_error(cfr_urls(year = good_year,
#                           title_number = bad_title_number,
#                           check_url = TRUE), "Title must be a numeric value between 1 and 50.\n")
#
#   testthat::expect_message(cfr_urls(year = na_year,
#                             title_number = na_title_number,
#                             verbose = TRUE,
#                             check_url = FALSE), sprintf("There aren't any regulations for title %s in %s.",
#                                                         na_title_number,
#                                                         na_year))
#
#   testthat::expect_true(is.na(cfr_urls(year = na_year,
#                          title_number = na_title_number,
#                          verbose = FALSE,
#                          check_url = FALSE)))
#
#     cfr <- cfr_urls(year = good_year,
#                     title_number = good_title_number,
#                     check_url = FALSE)
#
#     testthat::expect_true(all(grepl("*..xml$", cfr)))
#     testthat::expect_true(is.character(cfr))
#
# })
#
# test_that("We can parse some parts.", {
#
#   good_year <- 2000
#   good_title_number <- 15
#
#   bad_year <- 1993
#   bad_title_number <- 55
#
#   na_year <- 1996
#   na_title_number <- 1
#
#   na_url <- cfr_urls(year = na_year, title_number = na_title_number)
#   good_url <- cfr_urls(year = good_year, title_number = good_title_number)[1]
#   bad_url <- sprintf("https://www.gpo.gov/fdsys/bulkdata/CFR/%s/title-%s/CFR-%s-title%s-vol1.xml",
#                      bad_year,
#                      bad_title_number,
#                      bad_year,
#                      bad_title_number)
#
#   testthat::expect_error(cfr_part(bad_url), "The URL is not valid.")
#
#   testthat::expect_error(cfr_part(na_url), "NA is not a valid url.")
#
#   testthat::expect_message(cfr_part(good_url,
#                           verbose = TRUE),
#                  sprintf("Pulling the chapter, part, and volume information from:\n%s.\n",
#                          good_url))
#
#   good_part <- cfr_part(good_url)
#
#   testthat::expect_true(is.data.frame(good_part))
#   testthat::expect_true(good_url == unique(good_part$url))
#   testthat::expect_true(good_year == unique(good_part$year))
#   testthat::expect_true(good_title_number == unique(good_part$title))
#   testthat::expect_true(all(grepl("*..xml$", good_part$url)))
#
# })

test_that("We can extract some numbers.", {

  bad_part <- "800 to end"
  good_part <- "Part 800 to end"
  silly_part <- "Part no digits"

  testthat::expect_error(numextract(bad_part), "Make sure you are providing a valid 'part'.")
  testthat::expect_error(numextract(silly_part), "Make sure string is a numeric value.")

  testthat::expect_true(numextract("Part 100 to 200", return = "max") == 200)
  testthat::expect_true(numextract("Part 100 to 200", return = "min") == 100)
  testthat::expect_true(numextract("Part 100 to end", return = "max") == Inf)

  testthat::expect_true(is.numeric(numextract("Part 100 to 200", return = "max")))
  testthat::expect_true(is.numeric(numextract("Part 100 to 200", return = "min")))
  testthat::expect_true(is.numeric(numextract("Part 100 to end", return = "max")))

})


test_that("We can go all the way", {

  good_year <- 2012
  bad_year <- 1995
  na_year <- 1996
  good_title_number <- 50
  bad_title_number <- 1000
  na_title_number <- 1
  good_chapter <- 6
  bad_chapter <- "BB"
  good_part <- 648
  bad_part <- "DD"

  testthat::expect_error(cfr_text(bad_year,
                                  good_title_number,
                                  good_chapter,
                                  good_part),
                         "Year must be between 1996 and 2018.\n")

  testthat::expect_error(cfr_text(good_year,
                                  bad_title_number,
                                  good_chapter,
                                  good_part),
                         "Title must be a numeric value between 1 and 50.\n")

  testthat::expect_error(cfr_text(good_year,
                                  good_title_number,
                                  bad_chapter,
                                  good_part),
                         "Chapter must be a numeric value, not a Roman Numeral.\n")

  testthat::expect_error(cfr_text(good_year,
                                  good_title_number,
                                  good_chapter,
                                  bad_part),
                         "Part must be a numeric value.\n")

  testthat::expect_error(cfr_text(na_year,
                                  na_title_number,
                                  good_chapter,
                                  good_part),
                         sprintf("There aren't any regulations for title %s in %s.\n", na_title_number, na_year))


  good_text <- cfr_text(good_year, good_title_number, good_chapter, good_part)


  testthat::expect_true(all(class(good_text) %in% c("grouped_df", "tbl", "tbl_df", "data.frame") == TRUE))

})


R/functions.R
#' cfr_part
#'
#' @title Parse the Relevant Details for CFR xml files
#' @description \code{cfr_part} returns a data_frame year, title, volume, chapters, parts, and file path for each xml file
#' @details Since we're after more refined data than a single volume, we need to figure out what chapters and parts are associated with each volume. This function parses the xml and scrapes the Table of Contents for the information held in each volume.
#'
#' @param verbose logical. Will return "helpful" messages regarding the status of the file path.
#' @param file_path file path coming from the bulk download
#'
#' @return Numeric (year, title, volume, and chapters) and characters (parts and file path).
#' @importFrom magrittr %>%
#' @keywords internal
#'

cfr_part <- function(file_path, verbose = FALSE){

  res <- xml2::read_xml(file_path, as = "parsed", encoding = "UTF-8")

  parts <- res %>%
    xml2::xml_find_all(".//FMTR/TITLEPG/PARTS") %>%
    xml2::xml_text()

  chapters <- res %>%
    xml2::xml_find_all(".//FMTR/TOC/TITLENO/CHAPTI/SUBJECT") %>%
    xml2::xml_text()

  if(length(chapters) == 0){
    chapters <- res %>%
      xml2::xml_find_all(".//TOC/TITLENO/CHAPTI/SUBJECT") %>%
      xml2::xml_text()
  }

  if(verbose) {
    message(sprintf("Pulling the chapter, part, and volume information from:\n%s.\n", file_path))
  }

  return(data.frame(year = as.numeric(gsub(".*CFR-(.*)-title(.*)-vol(.*).xml", "\\1", file_path)),
                    title = as.numeric(gsub(".*CFR-(.*)-title(.*)-vol(.*).xml", "\\2", file_path)),
                    vol = as.numeric(gsub(".*CFR-(.*)-title(.*)-vol(.*).xml", "\\3", file_path)),
                    chapters = chapters,
                    parts = parts,
                    file_path = file_path,
                    stringsAsFactors = FALSE))
}


#' numextract
#'
#' @title Extract the Part Numbers
#' @description \code{numextract} takes the part numbers from \code{cfr_part} output.
#'
#' @details Each CFR chapter has multiple parts that often span volumes. To facilitate targeting a specific part, it's necessary to evaluate which parts are in each volume (e.g., Parts 18 to 199). The CFR sometimes uses terms like "END" or "end" to denote the maximum part in each chapter. \code{numextract} simply returns the max as `Inf` in these situations.
#'
#' @param string ideally \code{cfr_part()}$parts
#' @param return min or max, default is "min"
#'
#' @return numeric value from 1 to `Inf``
#'
#' @keywords internal
#'
numextract <- function(string, return = c("min", "max")[1]){

  if(!grepl("part*", tolower(string))){
    stop("Make sure you are providing a valid 'part'.")
  }

  num_vec <- unlist(regmatches(string,
                               gregexpr("[[:digit:]]+\\.*[[:digit:]]*", string)))

  if(length(num_vec) == 0) {
    stop("Make sure string is a numeric value.")
  }

  if(length(num_vec) == 1){
    num_vec <- c(num_vec, Inf)
  }
  if(return == "min"){
    return(as.numeric(num_vec[1]))
  }
  if(return == "max"){
    return(as.numeric(max(num_vec)))
  }
}

#' cfr_text
#'
#' @title Extract the Text for a Given Year, Title, Chapter, and Part
#' @description \code{cfr_text} returns a tibble of CFR text
#' @details This function is the main function of the \code{fedregs} package. It takes the title, chapter, part, and year and returns a tibble of raw text (\code{return_tidytext = FALSE}) or \href{https://www.tidytextmining.com/tidytext.html}{tidytext} text (\code{return_tidytext = TRUE}). N.b., it has not been extensively tested on titles and chapters other than Title 50 chapter VI and part 648.
#'
#' @param year numeric between 1996 and 2018.
#' @param title_number numeric between 1 and 50.
#' @param chapter numeric or roman numeral.
#' @param part numeric.
#' @param token character. Unit for tokenizing. Currently
#' @param return_tidytext logical. TRUE = tidytext, FALSE = raw data
#' @param verbose logical. Will return "helpful" messages regarding the status of the URL.
#' @param ... Extra arguments passed on to tokenizers, such as n and k for "ngrams" and "skip_ngrams"
#'
#' @return a tibble with year, title_number, chapter, part, and text nested by subpart
#' @export
#' @importFrom magrittr %>%
#' @importFrom utils download.file unzip
#' @importFrom tidyr nest unnest
#'
#' @examples
#' \donttest{regs <- cfr_text(year = 2018,
#' title_number = 50,
#' chapter = 6,
#' part = 648,
#' return_tidytext = TRUE,
#' token = "words",
#' verbose = TRUE)
#' head(regs)}
#'

cfr_text <- function(year, title_number, chapter, part, token = "words", return_tidytext = TRUE,
                     verbose = FALSE, ...) {

  if(!year %in% seq(1996, 2018)){
    stop("Year must be between 1996 and 2018.\n")
  }

  if(!title_number %in% seq(1, 50)){
    stop("Title must be a numeric value between 1 and 50.\n")
  }

  if(!is.numeric(chapter)){
    stop("Chapter must be a numeric value, not a Roman Numeral.\n")
  }
  if(is.numeric(chapter)){
    chapter <- as.character(utils::as.roman(chapter))
  }

  if(!is.numeric(part)){
    stop("Part must be a numeric value.\n")
  }

  if(token == "ngram" & !exists("n", mode = "integer")) {
    stop("For ngram tokens, please include the 'n' argument.")
  }


  url_head <- "https://www.govinfo.gov/bulkdata/CFR"
  url_zip <- sprintf("%s/%s/title-%s/CFR-%s-title-%s.zip", url_head, year, title_number, year, title_number)

  if(httr::http_error(url_zip)){
    stop(sprintf("There aren't any regulations for title %s in %s.\n", title_number, year))
  }

  temp_dir <- tempdir()
  temp <- tempfile(tmpdir = temp_dir)
  download.file(url_zip, temp, quiet = !verbose)

  ## Now check the file to find the right volume
  unzip(temp, exdir = temp_dir)
  paths <- grep("*.xml$", list.files(temp_dir, full.names = TRUE), value = TRUE)
  paths <- grep(sprintf("*CFR-%s-title%s", year, title_number), paths, value = TRUE)

  cfr_part_df <- purrr::map_df(paths, cfr_part, verbose = verbose)

  cfr_select_part <- cfr_part_df %>%
    dplyr::mutate(min_parts = purrr::map(parts, numextract, "min"),
                  max_parts = purrr::map(parts, numextract, "max")) %>%
    dplyr::filter(grepl(sprintf("Chapter %s", chapter), chapters),
                  min_parts <= part,
                  max_parts >= part) %>%
    tidyr::unnest(cols = c("min_parts", "max_parts"))


  if(nrow(cfr_select_part) == 0){
    stop(sprintf("Part not found. Please check that part %s is available on 'https://www.govinfo.gov/bulkdata/CFR'.\n", part))
  }

  max_parts_vector <- unlist(cfr_select_part$max_parts)

  if(nrow(cfr_select_part) > 1 &
    any(!is.infinite(max_parts_vector))) {
    cfr_select_part <- cfr_select_part %>%
      dplyr::filter(!is.infinite(max_parts))
  }

  if(nrow(cfr_select_part) > 1 &
     any(is.infinite(max_parts_vector))) {
    cfr_select_part <- cfr_select_part %>%
      dplyr::filter(min_parts == max(min_parts))
  }


  cfr_xml <- cfr_select_part %>%
    dplyr::select(c(file_path)) %>%
    dplyr::pull() %>%
    xml2::read_xml(as = "text", encoding = "UTF-8")

  subpart_names <- cfr_xml %>%
    xml2::xml_find_all(sprintf(".//PART/HD[contains(text(), '%s')]/following-sibling::CONTENTS/SUBPART/HD",
                               part)) %>%
    xml2::xml_text(.) %>%
    unlist()

  section_all <- dplyr::tibble(subpart_names) %>%
    dplyr::mutate(values = purrr::map(subpart_names, function(x) xml2::xml_find_all(x = cfr_xml,
                                                              sprintf(".//PART/HD[contains(text(), '%s')]/following-sibling::CONTENTS/SUBPART/HD[contains(text(), '%s')]/following-sibling::SECTNO",
                                                                      part,
                                                                      x)) %>%
                                 xml2::xml_text(.))) %>%
    tidyr::unnest(cols = c("values"))

  cfr_subpart <- cfr_xml %>%
    xml2::xml_find_all(sprintf(".//PART/HD[contains(text(), '%s')]/following-sibling::SUBPART", part))

  section_names <- cfr_subpart %>%
    xml2::xml_find_all(".//SECTION/SUBJECT|.//SECTION/RESERVED") %>%
    xml2::xml_text() %>%
    unlist()
  section_names <- section_names[!section_names == ""]

  section_numbers <-  cfr_subpart %>%
    xml2::xml_find_all(".//SECTION/SECTNO") %>%
    xml2::xml_text() %>%
    unlist()

  # section_numbers <- cfr_subpart %>%
  #   purrr::map(~ xml2::xml_find_all(., "//SECTION/SECTNO")) %>%
  #   purrr::map(~ xml2::xml_text(.)) %>%
  #   unlist()
  #
  # section_names <- cfr_subpart %>%
  #   purrr::map(~ xml2::xml_find_all(., "//SECTION/SUBJECT|//SECTION/RESERVED")) %>%
  #   purrr::map(~ xml2::xml_text(.)) %>%
  #   unlist()

  section_text <- dplyr::tibble(SECTION_NAME = section_names,
                                SECTION_NUMBER = section_numbers,
                                values = gsub(".*\u00A7|.*\u2009|.*\\s", "", section_numbers)) %>%
    dplyr::filter(grepl(sprintf("%s[\\.]", part), SECTION_NUMBER)) %>%
    dplyr::left_join(section_all, by = "values") %>%
    dplyr::mutate(TEXT = purrr::map(SECTION_NUMBER, function(x) xml2::xml_find_all(cfr_subpart,
                                                                            sprintf(".//SECTNO[text()='%s']/following-sibling::P",
                                                                                    x)) %>%
                               xml2::xml_text(., trim = TRUE)),
           year = year,
           title_number = title_number,
           chapter = chapter,
           part = part,
           TEXT = tolower(TEXT)) %>%
    dplyr::rename(subpart = subpart_names)


    if(return_tidytext){
    out <- section_text %>%
             tidytext::unnest_tokens(word, TEXT, token, ...) %>%
             dplyr::group_by(year, title_number, chapter, part, subpart) %>%
             tidyr::nest()
  }

  if(!return_tidytext){
    out <- section_text %>%
             dplyr::group_by(year, title_number, chapter, part, subpart) %>%
             tidyr::nest()
  }
  return(out)
}


analysis/analysis.R
devtools::install_github("slarge/fedregs@dev")
library(dplyr)
library(fedregs)

## Word counts
total_words <- function(year, title_number, chapter, part){

  tt <- cfr_text(year = year,
                 title_number = title_number,
                 chapter = chapter,
                 part = part,
                 return_tidytext = TRUE,
                 verbose = FALSE)

  stop_words <- dplyr::data_frame(word = quanteda::stopwords("english"))

  clean_words <- tt %>%
    tidyr::unnest() %>%
    dplyr::mutate(word = gsub("[[:punct:]]", "", word), # remove punctuation
                  word = gsub("^[[:digit:]]*", "", word)) %>%  # remove digits (e.g., 1st, 1881a, etc)
    dplyr::anti_join(stop_words, by = "word") %>%  # remove "stop words"
    dplyr::filter(is.na(as.numeric(word)),
                  !grepl("^m{0,4}(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$",
                         word),
                  !grepl("\\b[a-z]{1}\\b", word),
                  !grepl("\\bwww*.", word)) %>%
    dplyr::mutate(word = quanteda::tokens(word),
                  word = quanteda::tokens_wordstem(word),
                  word = as.character(word))

  count_words <- clean_words %>%
    summarise(year = unique(year),
              n = n())

  return(count_words)

}

# td <- lapply(2012:2013, function(x) total_words(year = x, title_number = 50, chapter = 6, part = 648))
# title_50 <- purrr::map_df(1996:2017, ~ cfr_text(year = .x,
#                title_number = 50,
#                chapter = 6,
#                part = 648,
#                return_tidytext = TRUE,
#                verbose = TRUE))
#
# saveRDS(title_50, "analysis/title50_1996-2017.RDS")
library(dplyr)
library(ggplot2)
title_50 <- readRDS("analysis/title50_1996-2017.RDS")

stop_words <- dplyr::data_frame(word = quanteda::stopwords("english"))

clean_words <- title_50 %>%
  tidyr::unnest() %>%
  dplyr::mutate(word = gsub("[[:punct:]]", "", word), # remove punctuation
                word = gsub("^[[:digit:]]*", "", word)) %>%  # remove digits (e.g., 1st, 1881a, etc)
  dplyr::anti_join(stop_words, by = "word") %>%  # remove "stop words"
  dplyr::filter(is.na(as.numeric(word)),
                !grepl("^m{0,4}(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$",
                       word),
                !grepl("\\b[a-z]{1}\\b", word),
                !grepl("\\bwww*.", word)) %>%
  dplyr::mutate(word = quanteda::tokens(word),
                word = quanteda::tokens_wordstem(word),
                word = as.character(word))

count_words <- clean_words %>%
  mutate(subpart_name = gsub("\u2014.*","", subpart)) %>%
  group_by(year, subpart_name) %>%
  summarise(n = n())

unique(clean_words$subpart)
library(ggplot2)
ggplot(count_words, aes(x = year, y = n))+
  geom_line(aes(color = subpart_name))+
  # facet_wrap(~subpart_name) +
  labs(xlab = NULL,
     title = "Code of Federal Regulations",
     subtitle = "Title 50, Chapter VI, Part 648",
     caption = sprintf("Data accessed on %s from:\n https://www.gpo.gov/fdsys/browse/collectionCfr.action?collectionCode=CFR",
                       format(Sys.Date(), "%d %B %Y"))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom",
        text = element_text(size = 8)) +
  theme_minimal()



title_50 <- purrr::map_df(1996:2017, ~ total_words(year = .x,
                                                   title_number = 50,
                                                   chapter = 6,
                                                   part = 648))





library(ggplot2)
ggplot(td, aes(x = year, y = n)) +
  geom_line() +
  theme_minimal()

# td will be a data.frame of year and total number of words



## #3 look at the bind_tf_idf
## work through: https://www.tidytextmining.com/tfidf.html#the-bind_tf_idf-function

year_words <- purrr::map_df(2016:2017, ~ cfr_text(year = .x,
                                              title_number = 50,
                                              chapter = 6,
                                              part = 648,
                                              return_tidytext = TRUE,
                                              verbose = FALSE))
# year_words <- bind_rows(cfr_text(year = 2015,
#                                  title_number = 50,
#                                  chapter = 6,
#                                  part = 648,
#                                  return_tidytext = TRUE,
#                                  verbose = FALSE),
#                         cfr_text(year = 2016,
#                                  title_number = 50,
#                                  chapter = 6,
#                                  part = 648,
#                                  return_tidytext = TRUE,
#                                  verbose = FALSE))


book_words <- clean_words %>%
  tidyr::unnest() %>%
  count(year, word, sort = TRUE)

book_words <- book_words %>%
  bind_tf_idf(word, year, n)
book_words


book_words %>%
  # select(-total) %>%
  arrange(desc(tf_idf))


book_words %>%
  arrange(desc(tf_idf)) %>%
  mutate(word = factor(word, levels = rev(unique(word)))) %>%
  group_by(year) %>%
  top_n(5) %>%
  ungroup %>%
  ggplot(aes(word, tf_idf, fill = year)) +
  geom_col(show.legend = FALSE) +
  labs(x = NULL, y = "tf-idf") +
  facet_wrap(~year, ncol = 2, scales = "free") +
  coord_flip()


part_vec <- cfr_urls(year = 2017, title_number = 50)
lapply(part_vec, cfr_part)


title_50_all <- purrr::map2(1996:2017, c(600, 622, 628,) ~ cfr_text(year = .x,
               title_number = 50,
               chapter = 6,
               part = 648,
               return_tidytext = TRUE,
               verbose = TRUE))

saveRDS(title_50, "analysis/title50_1996-2017.RDS")



analysis/reg_ratio.R
# devtools::install_github("slarge/fedregs")
library(dplyr)
library(fedregs)
# library(quanteda)
# library(tidyr)
library(ggplot2)

# reg <- purrr::map_df(1996:2017, ~ cfr_text(year = .x,
#                                           title_number = 50,
#                                           chapter = 6,
#                                           part = 648,
#                                           return_tidytext = FALSE,
#                                           verbose = TRUE))

# saveRDS(reg, "analysis/title50_1996-2017.rds")
reg <- readRDS("analysis/title50_1996-2017.rds")
bad_words <- c("to", "is", "a", "the", "that", "this", "it","its","itself",
               "these", "those", "am", "were", "are", "was", "and", "but", "an",
               "if", "or", "because", "as", "of", "at", "in", "for", "be", "by",
               "with", "on", "any", "than", "all", "been", "each", "such")

bad_words <- paste(bad_words, collapse = "|")

bad_words <- paste0("\\b(", bad_words, ")\\b")

stop_words <- quanteda::stopwords("english")

bad_words[bad_words %in% stop_words]
raw_words <- stop_words[!stop_words %in% bad_words]


td <- unique(word_regs$word[word_regs$word %in% raw_words])

clean_regs <- reg %>%
  tidyr::unnest() %>%
  mutate(TEXT = gsub("^c\\(", "", TEXT),
         TEXT = gsub("([/-])|[[:punct:]]", "\\1", TEXT), # remove any remaining punctuation
         TEXT = gsub("[[:digit:]]", "", TEXT),
         TEXT = gsub("—", "", TEXT),
         TEXT = gsub("[^[:ascii:]]", "", TEXT, perl = TRUE),
         # TEXT = gsub(bad_words, "", TEXT),
         NULL)

word_regs <- clean_regs %>%  # remove digits (e.g., 1st, 1881a, 15th, etc)
  mutate(TEXT = gsub("\\bmay not\\b", "weenieroast", TEXT)) %>%
  tidytext::unnest_tokens(output = word, input = TEXT, token = "words") %>%
  filter(is.na(as.numeric(word)),
         !grepl("^m{0,4}(cm|cd|d?c{0,3})(xc|xl|l?x{0,3})(ix|iv|v?i{0,3})$",
                word), # adios Roman Numerals
         !grepl("\\b[a-z]{1}\\b", word), # get rid of one letter words
         !grepl("\\bwww*.", word)) %>%
  dplyr::mutate(word = quanteda::char_wordstem(word),
                word = case_when(word == "weenieroast" ~ "may not",
                                 TRUE ~ word))

count_regs <- word_regs %>%
  group_by(year) %>%
  summarize(n = n())

reg_words <- c("shall", "must", "may not", "prohibited", "required")

ratio_regs <- word_regs %>%
  filter(word %in% reg_words) %>%
  group_by(year) %>%
  summarize(reg_words = n()) %>%
  left_join(count_regs, by = c("year")) %>%
  mutate(ratio = reg_words/n)



count_regs %>%
  arrange(-words) %>%
  head(100) %>%
  pull(word)





count_words <- clean_words %>%
  group_by(word) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  arrange(-n) %>%
  top_n(n = 25, wt = n) %>%
  mutate(word = reorder(word, n))

ggplot(ratio_regs, aes(x = year, ratio)) +
  geom_line() +
  labs(xlab = NULL,
       title = "Code of Federal Regulations (2017)",
       subtitle = "Title 50, Chapter VI, Part 648",
       caption = sprintf("Data accessed on %s from:\n https://www.gpo.gov/fdsys/browse/collectionCfr.action?collectionCode=CFR",
                         format(Sys.Date(), "%d %B %Y"))) +
  # theme(axis.text.x = element_text(angle = 45, hjust = 1),
  #       legend.direction = "horizontal",
  #       legend.position = "bottom",
  #       text = element_text(size = 8)) +
  # coord_flip() +
  theme_minimal()
